swift-sdk/Internal/EmbeddedSessionManager.swift

```diff
@@ -0,0 +1,86 @@
+//
+//  EmbeddedSessionManager.swift
+//  
+//
+//  Created by Andrew Nelson on 5/30/23.
+//
+
+import Foundation
+
+public class EmbeddedSessionManager {
+    public static let shared = EmbeddedSessionManager()
+
+    public var currentSession: IterableEmbeddedSession?
+    private var messageTimers: [String: Timer] = [:]
+    var pauseTimestamps: [String: Date] = [:]
+
+
+    public func startSession() {
+        let startTime = Date()
+        currentSession = IterableEmbeddedSession(embeddedSessionId: UUID().uuidString, embeddedSessionStart: startTime, impressions: [])
+    }
+
+    public func endSession() {
+        currentSession?.embeddedSessionEnd = Date()
+        for timer in messageTimers.values {
+            timer.invalidate()
+        }
+        messageTimers.removeAll()
+        
+        if let session = currentSession {
+            let _ = IterableAPI.embeddedMessagingManager.track(embeddedSession: session)
+        }
+    }
+
+    public func stopTimerForImpression(impressionId: String) {
+        messageTimers[impressionId]?.invalidate()
+        messageTimers[impressionId] = nil
+    }
+    
+    public func pauseTimerForImpression(impressionId: String) {
+        messageTimers[impressionId]?.invalidate()
+        pauseTimestamps[impressionId] = Date()
+    }
+
+    public func resumeTimerForImpression(impressionId: String) {
+        if let timer = messageTimers[impressionId], timer.isValid {
+            return
+        }
+        if pauseTimestamps.removeValue(forKey: impressionId) != nil {
+            if let index = currentSession?.impressions.firstIndex(where: { $0.messageId == impressionId }) {
+                let newDisplayDuration = currentSession!.impressions[index].displayDuration
+                currentSession?.impressions[index].displayDuration = newDisplayDuration
+            }
+            let timer = Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { [weak self] _ in
+                self?.updateDisplayDurationForImpression(impressionId: impressionId)
```

Comment:
Hmm so theres a method called every one second for every on going impression? what if resume and pause were actually just flags and when pausing, resuming and ending an impression we just calculate the time difference and add/update and store them?

Reply:
That's a great idea. I'm refactoring the class to use another time approach via timestamps and flags